rust/template/d3log/src/record_set.rs

```diff
@@ -0,0 +1,125 @@
+// functions to allow a set of Records, a dynamically typed alternative to DDValue, to act as
+// Batch for interchange between different ddlog programs
+
+use crate::{Batch, BatchBody, Error};
+use differential_datalog::record::Record;
+use std::collections::HashMap;
+use std::fmt;
+use std::fmt::Display;
+
+#[derive(Clone, Default)]
+pub struct RecordSet {
+    pub records: Vec<(Record, isize)>,
+}
+
+// it turns out to be useful to use fully qualified names in macros
+// to avoid asking everyone else to set up their namespaces properly
+#[macro_export]
+macro_rules! fact {
+    ( $rel:path,  $($n:ident => $v:expr),* ) => {
+        Batch{body: BatchBody::Record(RecordSet::singleton(
+            Record::NamedStruct(
+                Cow::from(stringify!($rel).to_string()),
+                vec![$((Cow::from(stringify!($n)), $v),)*]))),
+              metadata: std::collections::HashMap::new(),
+        }
+    }
+}
+
+impl Display for RecordSet {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut m = HashMap::new();
+        for (r, _w) in self.records.clone() {
+            match r {
+                Record::Bool(_b) => println!("unprintable record type bool"),
+                Record::Int(_i) => println!("unprintable record type int"),
+                Record::Float(_f) => println!("unprintable record type float"),
+                Record::Double(_dbl) => println!("unprintable record type double"),
+                Record::String(string_name) => println!("{}", string_name),
+                Record::Serialized(name, _s) => println!("serialized {}", name),
+                Record::Tuple(t) => {
+                    println!("tuple {:?}", t);
+                }
+                Record::Array(_, _record_vec) => println!("unprintable record type array"),
+                Record::PosStruct(name, _record_vec) => println!("{}", name),
```

Comment:
Why are you using `println` within a `Display` implementation? Additionally, I'm not sure what about these is "unprintable"